### PR TITLE
Nessie: Use unique path for different table with same name

### PR DESCRIPTION
**Pull request by ajantha-bhat**
In Nessie catalog, 
Each branch can have a new table with the same name as a table in another branch. When this happens, Nessie is using the same table path two different tables (when same warehouse location is configured).

So, use the table UUID in the table path to avoid two different tables sharing the same path. 

Steps:

```
docker run -p 19120:19120 projectnessie/nessie

                              
 /Users/ajantha/Documents/tools/spark-3.0.3-bin-hadoop2.7/bin/spark-sql --packages org.slf4j:slf4j-api:1.7.32 --jars /Users/ajantha/Downloads/iceberg-spark-runtime-3.0_2.12-0.14.0-SNAPSHOT.jar,/Users/ajantha/Downloads/nessie-spark-extensions-0.28.0.jar \
                                    --conf spark.sql.catalog.nessie=org.apache.iceberg.spark.SparkCatalog\
                                    --conf spark.sql.catalog.nessie.warehouse=$PWD/nessie2 \
                                --conf spark.sql.catalog.nessie.catalog-impl=org.apache.iceberg.nessie.NessieCatalog \
                                --conf spark.sql.catalog.nessie.uri=http://localhost:19120/api/v1 \
                                --conf spark.sql.catalog.nessie.ref=main \
                                --conf spark.sql.catalog.nessie.auth-type=NONE \
                                --conf spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions,org.projectnessie.spark.extensions.NessieSparkSessionExtensions
```

```
create branch dev in nessie from main;

use reference dev in nessie;

create table nessie.db1.t1(id int) using iceberg;
insert into nessie.db1.t1 select 42;

use reference main in nessie;

create table nessie.db1.t1(place string, name string) using iceberg;
insert into nessie.db1.t1 select 'x','y';
```
-- Now the path `$warehouse/db1/t1` has data and metadata files from both the tables.

**Why this needs to be fixed ?**
`remove_orphan` (Also Nessie GC tool) files does a list files (`BaseDeleteOrphanFilesSparkAction#buildActualFileDF()` -> `listDirsRecursively`) and try to remove the files that are not reachable from metadata. So, it will clean up other table in this case.

Not sure if there are any other Iceberg operations that do list files. If it is there. It will get impacted too.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 1]
```diff
@@ -157,19 +161,27 @@ public TableBuilder withProperty(String key, String value) {
       return this;
     }
 
+    public TableMetadata newTableMetadata() {
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:164]: Extracting the logic that needs to be overridden for Nessie. 

Without changing this class, need to copy-paste a lot of code from here to NessieCatalog for overriding. So, code duplication will be more.

**Comment by ajantha-bhat** [comment]
cc: @nastra, @snazy, @RussellSpitzer, @rdblue

**Comment by ajantha-bhat** [comment]
ping...

File: nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java [hunk 2]
```diff
@@ -443,4 +443,26 @@ public void testWithRefAndHash() throws NessieConflictException, NessieNotFoundE
...
+    String branch1 = "branch1";
+    String branch2 = "branch2";
+    createBranch(branch1, null);
+    createBranch(branch2, null);
+    Schema schema1 = new Schema(Types.StructType.of(required(1, "id", Types.LongType.get())).fields());
+    Schema schema2 = new Schema(Types.StructType.of(
+        required(1, "file_count", Types.IntegerType.get()),
+        required(2, "record_count", Types.LongType.get())).fields());
+
+    TableIdentifier identifier = TableIdentifier.of("db1.", "table1");
```
Comment (nastra) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java:458]: nit: replace `db1.` with `db`

File: nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java [hunk 3]
```diff
@@ -443,4 +443,26 @@ public void testWithRefAndHash() throws NessieConflictException, NessieNotFoundE
...
+
+    NessieCatalog nessieCatalog = initCatalog(branch1);
+    String location1 = nessieCatalog.createTable(identifier, schema1).location();
```
Comment (nastra) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java:461]: can we also verify that each table has the correct schema?

**Review by nastra — COMMENTED** [review]
in general this LGTM, just a few things

File: nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java (hunk 3 above)
Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java]: I didn't add it initially as it is a base functionality and it is already validated in some other test suite.

added now.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java (hunk 1 above)
Comment (snazy) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:164]: Should be `protected`

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 4]
```diff
@@ -157,19 +161,27 @@ public TableBuilder withProperty(String key, String value) {
...
+      String baseLocation = location != null ? location : defaultWarehouseLocation(identifier);
+      tableProperties.putAll(tableOverrideProperties());
+      return TableMetadata.newTableMetadata(schema, spec, sortOrder, baseLocation, tableProperties);
+    }
+
+    public TableMetadata replacedTableMetadata(TableOperations ops) {
```
Comment (snazy) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:170]: Should be `protected`

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java [hunk 5]
```diff
@@ -256,4 +263,31 @@ String currentRefName() {
...
+    }
+
+    private TableMetadata updateLocation(TableMetadata tableMetadata) {
+      if (location() != null) {
+        // create the table at the location specified.
+        return tableMetadata;
+      }
+      // Different tables with same table name can exist across references in Nessie.
+      // To avoid sharing same table path between two tables with same name, use table uuid in the table path.
+      return tableMetadata.updateLocation(tableMetadata.location() + "_" + tableMetadata.uuid());
```
Comment (snazy) [inline nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:290]: I'd really prefer a new, random UUID instead of the table's UUID.

Comment (ajantha-bhat) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java]: ok. Used the random UUID instead of table UUID.

**Comment by ajantha-bhat** [comment]
@snazy , @nastra : Addressed the comments. Please have a look at it again.

**Review by nastra — APPROVED** [review]

**Review by snazy — APPROVED** [review]

**Comment by nastra** [comment]
@RussellSpitzer could you review/merge this one please?

**Comment by rdblue** [comment]
@ajantha-bhat, do you think Nessie could move to the REST catalog API? That API fixes several problems, like the one here. The table location should be something that the catalog endpoint can override, rather than requiring the client to produce the correct one.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java [hunk 6]
```diff
@@ -157,19 +161,21 @@ public TableBuilder withProperty(String key, String value) {
       return this;
     }
 
+    protected TableMetadata newTableMetadata() {
+      String baseLocation = location != null ? location : defaultWarehouseLocation(identifier);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:165]: @ajantha-bhat, why not override `defaultWarehouseLocation`? That is already intended to let the catalog customize the table locations.

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:165]: @rdblue:  Yes, overriding  `defaultWarehouseLocation` can solve the problem of two different tables with same identifier getting the same path (issue which this PR is trying to solve)

But `defaultWarehouseLocation()` is the public interface of the catalog. 
If user calls it `defaultWarehouseLocation()` N times for a table, then it gives N different location. Which is confusing.
 
So, Instead of overrding it, I fixed as described in the PR.

**Comment by ajantha-bhat** [comment]
> @ajantha-bhat, do you think Nessie could move to the REST catalog API? That API fixes several problems, like the one here. The table location should be something that the catalog endpoint can override, rather than requiring the client to produce the correct one.

I am not sure at the moment. I believe REST catalog is a client for some SERVER that implements the endpoints for these REST catalog calls. Nessie already has its own end points which are not same as what REST catalog expects. I think it is too late to change Nessie endpoints as it will break the compatibility. I might be wrong here. Please correct me If I am mistaken.

**Comment by ajantha-bhat** [comment]
I rebased and resolved the conflict.

File: core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java (hunk 6 above)
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:165]: Why would we be calling default warehouse location multiple times for the same table? Sounds like we would be safe here saying that "defaultWarehouseLocation" is not deterministic.

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:165]: Because it is a user-facing API,  (nothing stops the user from calling in two places while writing custom logic around it)
I felt it looks odd to return a different result for the same input.

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/BaseMetastoreCatalog.java:165]: @RussellSpitzer , @rdblue : I have replied above that why I didn't modify `defaultWarehouseLocation() `, If you still feel it is better to override `defaultWarehouseLocation() ` please let me know. I just want this issue to be fixed. Thanks.

**Comment by dimas-b** [comment]
I'm a bit out-of-date of the REST catalog impl., but could it have a common base class with the NessieCatalog perhaps?.. just an idea (for a future change)

**Comment by ajantha-bhat** [comment]
@rdblue , @RussellSpitzer : As there was no response, I assumed people are more inclined toward modifying the default warehouse location.  I have modified it as the same. Please help in merging this issue fix.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 7]
```diff
@@ -400,7 +405,7 @@ private void validateRegister(TableIdentifier identifier, String metadataVersion
 
   @Test
   public void testRegisterTableWithGivenBranch() {
-    List<String> metadataVersionFiles = metadataVersionFiles(TABLE_NAME);
+    List<String> metadataVersionFiles = metadataVersionFiles(tableNameWithUUID(catalog.loadTable(TABLE_IDENTIFIER)));
```
Comment (rdblue) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:408]: It seems odd to use the location to get a directory name, but then add the rest of the location back to list files. Why not just update `metadataVersionFiles` and `manifestFiles` to accept a full path?

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java [hunk 8]
```diff
@@ -181,10 +182,15 @@ protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
...
     if (table.hasNamespace()) {
-      return SLASH.join(warehouseLocation, table.namespace().toString(), table.name());
+      location = SLASH.join(warehouseLocation, table.namespace().toString(), table.name());
+    } else {
+      location = SLASH.join(warehouseLocation, table.name());
     }
-    return SLASH.join(warehouseLocation, table.name());
+    // Different tables with same table name can exist across references in Nessie.
+    // To avoid sharing same table path between two tables with same name, use uuid in the table path.
+    return location + "_" + UUID.randomUUID();
```
Comment (rdblue) [inline nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:193]: This looks reasonable to me.

**Comment by rdblue** [comment]
> I'm a bit out-of-date of the REST catalog impl., but could it have a common base class with the NessieCatalog perhaps?.. just an idea (for a future change)

I think what makes the most sense is to eventually use the same protocol, especially since the REST protocol solves a number of these challenges (like delegating location defaults to the service). It also uses a diff-based approach that can avoid a lot more conflicts. I'm not sure how easy it would be for Nessie to update to be compatible with that client API, though. And I imagine it isn't a high priority just to avoid commit conflicts and enable metadata caching.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 7 above)
Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java]: ok. Updated now.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 9]
```diff
@@ -531,27 +531,26 @@ protected String getTableLocation(String tableName) {
...
 
-  protected List<String> metadataVersionFiles(String tableName) {
-    return filterByExtension(tableName, getFileExtension(TableMetadataParser.Codec.NONE));
+  protected List<String> metadataVersionFiles(Path tablePath) {
+    return filterByExtension(tablePath.toString(), getFileExtension(TableMetadataParser.Codec.NONE));
   }
 
-  protected List<String> manifestFiles(String tableName) {
-    return filterByExtension(tableName, ".avro");
+  protected List<String> manifestFiles(Path tablePath) {
```
Comment (rdblue) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:548]: Why is this passing a `Path`, only to convert to string and strip the `file:` from the `toString` representation in `filterByExtension`? Can you just pass `Path`?

Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:545]: ok. Avoided `toString` conversion and  stripping the `file:`

**Comment by ajantha-bhat** [comment]
@rdblue: I have squashed the commits and rebased because of the spotless apply.

**Comment by ajantha-bhat** [comment]
@rdblue , @RussellSpitzer : ping...

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 10]
```diff
@@ -561,28 +561,26 @@ protected String getTableLocation(String tableName) {
...
-    return Paths.get(getTableBasePath(tableName), "metadata").toString();
-  }
-
   @SuppressWarnings(
       "RegexpSinglelineJava") // respecting this rule requires a lot more lines of code
-  private List<String> metadataFiles(String tableName) {
-    return Arrays.stream(Objects.requireNonNull(new File(metadataLocation(tableName)).listFiles()))
+  private List<String> metadataFiles(String tablePath) {
+    return Arrays.stream(
+            Objects.requireNonNull(new File((tablePath + "/" + "metadata")).listFiles()))
```
Comment (rdblue) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:568]: I'd rather not mix file system APIs. This uses both Path and File. Can you get rid of `Path` and make that a string everywhere?

Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:568]: I believe the original author of this file might referred this from `HiveTableBaseTest` which uses both `Path` and `File`

Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java]: @rdblue : Removed `Path` from the whole Nessie testsuite and using `File`

we can even do this for other modules like `HiveTableBaseTest` (but may be in a follow-up)

File: nessie/src/test/java/org/apache/iceberg/nessie/TestCustomNessieClient.java [hunk 11]
```diff
@@ -46,7 +46,7 @@ public void testNoCustomClient() {
         "nessie",
         ImmutableMap.of(
             CatalogProperties.WAREHOUSE_LOCATION,
-            temp.toUri().toString(),
```
Comment (rdblue) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestCustomNessieClient.java:49]: This was correct before. Location should be a full URI including the scheme.

Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestCustomNessieClient.java:49]: ok. I have updated now.  I hope it is ok now.

**Comment by ajantha-bhat** [comment]
ping @rdblue

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 12]
```diff
@@ -567,40 +565,29 @@ public void testGCEnabled() {
...
-    return new Path("file", null, Paths.get(getTableBasePath(tableName)).toString());
-  }
-
-  protected String getTableLocation(String tableName) {
-    return getTableLocationPath(tableName).toString();
-  }
-
-  private String metadataLocation(String tableName) {
-    return Paths.get(getTableBasePath(tableName), "metadata").toString();
+    return temp.toURI() + DB_NAME + "/" + tableName;
```
Comment (rdblue) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:568]: Does this automatically call `toString` on the URI from `temp`?

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 13]
```diff
@@ -567,40 +565,29 @@ public void testGCEnabled() {
...
   }
 
   @SuppressWarnings(
       "RegexpSinglelineJava") // respecting this rule requires a lot more lines of code
-  private List<String> metadataFiles(String tableName) {
-    return Arrays.stream(Objects.requireNonNull(new File(metadataLocation(tableName)).listFiles()))
+  private List<String> metadataFiles(File tablePath) {
+    return Arrays.stream(
+            Objects.requireNonNull(
+                new File((tablePath + "/" + "metadata").replace("file:", "")).listFiles()))
```
Comment (rdblue) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:576]: This looks like a safe way to work with strings, but not with files. It looks like you're still fighting the representation used everywhere. Here, you convert file to string, do custom manipulation, and go back to file. That seems needlessly complicated and has implications for readability and correctness (what does `new File(...)` return that requires `"file:"` to need removed??).

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 14]
```diff
@@ -92,15 +90,15 @@ public TestNessieTable() {
   @BeforeEach
   public void beforeEach(@NessieUri URI uri) throws IOException {
     super.beforeEach(uri);
-    this.tableLocation = new Path(catalog.createTable(TABLE_IDENTIFIER, schema).location());
+    this.tableLocation = new File(catalog.createTable(TABLE_IDENTIFIER, schema).location());
```
Comment (rdblue) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:93]: I don't really see why this isn't a `String` that gets converted to `File` when needed.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 12 above)
Comment (dimas-b) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:568]: Yes, apparently this is what it compiles to:
`javap -cp nessie/build/classes/java/test/ -p -c -s org.apache.iceberg.nessie.TestNessieTable`

```
  private java.lang.String getTableBasePath(java.lang.String);
    descriptor: (Ljava/lang/String;)Ljava/lang/String;
    Code:
       0: new           #88                 // class java/lang/StringBuilder
       3: dup
       4: invokespecial #89                 // Method java/lang/StringBuilder."<init>":()V
       7: aload_0
       8: getfield      #90                 // Field temp:Ljava/io/File;
      11: invokevirtual #91                 // Method java/io/File.toURI:()Ljava/net/URI;
      14: invokevirtual #92                 // Method java/lang/StringBuilder.append:(Ljava/lang/Object;)Ljava/lang/StringBuilder;
      17: ldc           #93                 // String db
      19: invokevirtual #94                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      22: ldc           #95                 // String /
      24: invokevirtual #94                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      27: aload_1
      28: invokevirtual #94                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      31: invokevirtual #96                 // Method java/lang/StringBuilder.toString:()Ljava/lang/String;
      34: areturn
```

but it's different with java 11, something like:
```
24: invokedynamic #9,  0              // InvokeDynamic #0:makeConcatWithConstants:(Ljava/net/URI;)Ljava/lang/String;
```

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 14 above)
Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java]: Hmm, I just followed the original author's style instead of refactoring it. 

Fixed now.

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 13 above)
Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java]: When `Table.location()` is called, it returns a string with `file:` scheme. So, while creating a `new File()` for a `ListFiles()`, if the `pathname` has the file scheme, `ListFiles` return null. Hence, removing the file scheme. Similar to how it is in many existing test cases. 

But instead of doing for every time, now I moved this logic to the `tablePath` initialization

**Comment by ajantha-bhat** [comment]
@gaborkaszab: I would like to get this one in 1.1 release.

File: nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java [hunk 15]
```diff
@@ -80,7 +80,7 @@ public abstract class BaseTestIceberg {
 
   private static final Logger LOG = LoggerFactory.getLogger(BaseTestIceberg.class);
 
-  @TempDir public Path temp;
+  @TempDir public File temp;
```
Comment (nastra) [inline nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:83]: what's the reason from switching from `Path` to `File`? The `File` API is considered **old** and `Path` should generally be preferred imo. Also the tests should work without having to change this. 
The only place you'd have to adjust is `TestNessieTable#testCreate()`: `String expected = (temp.toUri() + DB_NAME + "/" + tableName).replace("///", "/")` to `String expected = (temp.toUri() + DB_NAME + "/" + tableName)`

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java [hunk 16]
```diff
@@ -567,40 +567,29 @@ public void testGCEnabled() {
...
-    return Paths.get(getTableBasePath(tableName), "metadata").toString();
+    return temp.toURI() + DB_NAME + "/" + tableName;
   }
 
   @SuppressWarnings(
       "RegexpSinglelineJava") // respecting this rule requires a lot more lines of code
-  private List<String> metadataFiles(String tableName) {
-    return Arrays.stream(Objects.requireNonNull(new File(metadataLocation(tableName)).listFiles()))
+  private List<String> metadataFiles(String tablePath) {
+
```
Comment (nastra) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:576]: nit: unnecessary whitespace

File: nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java (hunk 15 above)
Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:83]: > https://github.com/apache/iceberg/pull/4826#discussion_r937079863
I'd rather not mix file system APIs. This uses both Path and File. Can you get rid of Path and make that a string everywhere?

Removed `Path` after this comment.

Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:83]: yeah, `Path` is more advanced but we don't need any specific function from `Path` that is not present in `File`.

Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:83]: > The only place you'd have to adjust

previously `metadataFiles`, `metadataVersionFiles`, `manifestFiles`, `metadataLocation` used to work with `tableName`, now because of uuid addition, it has to be location based. Hence, we started modifying the test cases and ended up here.

Comment (RussellSpitzer) [reply nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:83]: I agree that the change here probably isn't necessary. I would probably revert this change

File: nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java [hunk 17]
```diff
@@ -484,4 +484,32 @@ public void testWithRefAndHash() throws NessieConflictException, NessieNotFoundE
...
+    Schema schema1 =
+        new Schema(Types.StructType.of(required(1, "id", Types.LongType.get())).fields());
+    Schema schema2 =
+        new Schema(
+            Types.StructType.of(
+                    required(1, "file_count", Types.IntegerType.get()),
+                    required(2, "record_count", Types.LongType.get()))
+                .fields());
+
+    TableIdentifier identifier = TableIdentifier.of("db.", "table1");
```
Comment (RussellSpitzer) [inline nessie/src/test/java/org/apache/iceberg/nessie/TestBranchVisibility.java:503]: nit: why "db." ?

File: nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java (hunk 10 above)
Comment (RussellSpitzer) [reply nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java:578]: I would probably reserve larger refactors like this to another PR, not sure they are required to avoid usage of file here?

File: nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java (hunk 15 above)
Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:83]: > what's the reason for switching from Path to File?

If I switch to `Path` completely,  that also needs many changes as `TestNessieTable` validations/asserts has many  
`Assertions.assertThat(new File(foo)).exists();`
Also, some functions like `metadataFiles` use `listFiles()` that needs to be changed to `Files.walk()`.

**Review by dimas-b — COMMENTED** [review]
LGTM overall with minor comments.

The renaming suggestion for `defaultWarehouseLocation` is really optional. I'd appreciate some clean-up in test code though.

**Review by dimas-b — APPROVED** [review]

File: nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java (hunk 15 above)
Comment (nastra) [reply nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:83]: to me it seems that the switch from `Path` to `File` in this PR isn't required. I actually checked out the branch and reverted that change and everything was working after adjusting `TestNessieTable#testCreate()`

Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:83]: Do you mean, just with the **first commit** of this PR?
If different, can you share the branch with me?

Comment (nastra) [reply nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java:83]: I mean in this PR. The diff that is being done in this file isn't required imo

Comment (ajantha-bhat) [reply nessie/src/test/java/org/apache/iceberg/nessie/BaseTestIceberg.java]: Oh, you mean just this file. I have reverted now.

**Review by nastra — APPROVED** [review]

**Comment by ajantha-bhat** [comment]
@RussellSpitzer , @nastra : I have reverted the base test class changes and handled the nit.

File: nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java [hunk 18]
```diff
@@ -199,10 +200,16 @@ protected TableOperations newTableOps(TableIdentifier tableIdentifier) {
...
-      return SLASH.join(warehouseLocation, table.namespace().toString(), table.name());
+      location = SLASH.join(warehouseLocation, table.namespace().toString(), table.name());
+    } else {
+      location = SLASH.join(warehouseLocation, table.name());
     }
-    return SLASH.join(warehouseLocation, table.name());
+    // Different tables with same table name can exist across references in Nessie.
+    // To avoid sharing same table path between two tables with same name, use uuid in the table
+    // path.
+    return location + "_" + UUID.randomUUID();
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply nessie/src/main/java/org/apache/iceberg/nessie/NessieCatalog.java:212]: `defaultWarehouseLocation` -> my initial argument was same with Ryan. Later I realised, it is called only once.

**Comment by ajantha-bhat** [comment]
@RussellSpitzer, @rdblue: Can we please merge this PR if it is ok?

**Review by RussellSpitzer — APPROVED** [review]
I know our contract on defaultWarehouseLocation is a bit iffy here, but I agree this is what was discussed as the proper solution previously. If there are no further complaints I think we can merge this as is and reconsider the contract for defaultWarehouse or having an ancillary method in another PR.

**Comment by ajantha-bhat** [comment]
@RussellSpitzer: Thanks for the review and approval.

I think we can merge this as there is already a consensus for the fix (from everyone). 
The confusion and discussions were mainly about base test case refactoring. This can be done in follow-up too if there is any concern.  

@gaborkaszab: Can we please add `1.1.0 milestone` label for this?
